### ISSUE TITLE
https support not working

### ISSUE THREAD

**Issue by asuh** [chuckhoupt/awstats-dh#1]
This is a great project for Dreamhost, thanks for doing this!

After the update, I noticed my statistics were for only HTTP versions of the sites and not HTTPS. I'm looking through the for loop in `update-awstats` but it appears to be able to read through the https directories, although I don't know how to program in Bash.

**Comment by chuckhoupt** [comment]
Thanks for testing it out! The update script looks for the file `~/logs/example.com/https/access.log` to determine if `example.com` is active on HTTPS. Check to see if that file exists. Each domain's log directory should have `http` and `https` symbolic links. For example, one should see:

```
$ ls -F ~/logs/example.com/
http@  ... https@ ....
```

Occasionally, I've seen those symbolic links "go missing" and I've had to contact support to have it fixed.

Note that in the AWStats site list, HTTPS sites are unmarked (i.e. listed as "example.com"), while HTTP sites have a "-http" suffix.

**Comment by asuh** [comment]
Thank you, and yes both of these links do exist on the sites that use Let's Encrypt.

In the site list, I only see sites that contains "-http" suffix, nothing that's unmarked.

**Comment by chuckhoupt** [comment]
Ok, let's try stepping through the process to see where the problem starts. Try the following shell command.

The `update-awstats` script generates AWStats config files based on available `access.log` files. It uses the follow file-matching pattern to get the list:

```
$ echo $HOME/logs/*/http{,s}/access.log
/home/user/logs/example.com/http/access.log
/home/user/logs/example.com/https/access.log
...
```

The `access.log` files should contain readable log data. I.e.:

```
$ cat /home/user/logs/example.com/https/access.log
93.184.216.34  - - [04/Jan/2019:01:33:25 -0800] "GET / HTTP/1.1" 200 3227 "-" "..."
... 
```

The generated config files are stored in the `cgi-bin` dir, so you should see both marked/unmarked in there:

```
$ ls awstats-dh/cgi-bin/
awstats.example.com-http.conf
awstats.example.com.conf
...
```

The script then runs AWStats to produce data files in the `data` dir, so you should see both unmarked and marked data files. For example:

```
$ ls awstats-dh/data/
awstats012019.example.com-http.txt
awstats012019.example.com.txt
...
```

Finally, the awstats-dh web page lists sites based on the above data files, so at a minimum the unmarked data files must be missing.

**Comment by asuh** [comment]
The `echo` and `cat` commands return similar to what you described above. 

When I use `ls` into the `cgi-bin` and `data` dirs, it's all suffixed `-http` and no unmarked.

**Comment by chuckhoupt** [comment]
Ok, so it looks like the `update-awstats` script is stumbling on creating the config files. Could you run `update-awstats` and examine its output? It runs in trace-mode, so every command it runs is logged to stderr (with a + in front). Probably the last few lines will have a failing command and error message (if we're lucky).

**Comment by asuh** [comment]
Honestly, I don't know about `stderr` other than what I just read, which might be you're referring to the error log? No output in the log after running the command. If there's another way to read more errors inside the awstats dir, I need more info, thanks

**Comment by chuckhoupt** [comment]
Here's an example of running `update-awstats` in the shell/terminal, and the typical output it produces (first few lines, plus last two lines). If you have many sites, the output can be quite long.

The example below shows a successful run (ending with `+ cat`, `[[ -f ... ]]`). If it runs into an error, the last few lines will show the last commands executed and maybe an error message or other clues to the problem.

```
$ ~/example.com/awstats-dh/update-awstats

+ shopt -s nullglob
++ dirname ./update-awstats
+ cd .
++ pwd
+ BASE=/home/user/example.com/awstats-dh
+ CONFIGDIR=/home/user/example.com/awstats-dh/cgi-bin

 ... lots of lines as it loops thru site logs ...

+ cat
+ [[ -f /home/user/example.com/awstats-dh/awstats.example.com.conf ]]

$
```

**Comment by asuh** [comment]
Hmm, now that I see this, is it possible that there's a a problem with finishing the output. Is there a limit to how many log directories to process? I just looked at my log directory and I have a bunch of old websites in there that are not associated with the account anymore. 

It appears to stop processing after the first 35 directories, all of which only process HTTP versions. Does this do two rounds of processing and the 2nd one would be all the HTTPS? It's confusing why this wouldn't process HTTP and HTTPS at the same time.

For one domain in question that uses HTTPS, this is the output:

```
+ for LOG in '$HOME/logs/*/http{,s}/access.log'
+ [[ /home/user/logs/example.com/http/access.log =~ logs/(.+)/(http|https) ]]
+ SITE=example.com
+ PROTOCOL=http
+ case $PROTOCOL in
+ CONFIG=example.com-http
+ HTTPSLINKS=
+ CONFIG_PATH=/home/user/example.com/awstats-dh/cgi-bin/awstats.example.com-http.conf
+ LOCAL_CONFIG_PATH=/home/user/example.com/awstats-dh/awstats.example.com-http.conf
+ LOGALLCMD='zcat -f /home/user/logs/example.com/http/access.log.2* /home/user/logs/example.com/http/access.log |'
+ LOGDAYCMD='zcat -f /home/user/logs/example.com/http/access.log.0 /home/user/logs/example.com/http/access.log |'
+ case ${1:-all} in
+ write_config 'zcat -f /home/user/logs/example.com/http/access.log.2* /home/user/logs/example.com/http/access.log |'
+ cat
+ [[ -f /home/user/example.com/awstats-dh/awstats.example.com-http.conf ]]
+ /home/user/example.com/awstats-dh/awstats/wwwroot/cgi-bin/awstats.pl -configdir=/home/user/example.com/awstats-dh/cgi-bin -config=example.com-http -update
Create/Update database for config "/home/user/example.com/awstats-dh/cgi-bin/awstats.example.com-http.conf" by AWStats version 7.7 (build 20161204)
From data in log file "zcat -f /home/user/logs/example.com/http/access.log.2* /home/user/logs/example.com/http/access.log |"...
Phase 1 : First bypass old records, searching new record...
Searching new records from beginning of log file...
Phase 2 : Now process new records (Flush history on disk after 20000 hosts)...
Jumped lines in file: 0
Parsed lines in file: 3772
 Found 0 dropped records,
 Found 0 comments,
 Found 0 blank records,
 Found 0 corrupted records,
 Found 3739 old records,
 Found 33 new qualified records.
+ write_config 'zcat -f /home/user/logs/example.com/http/access.log.2* /home/user/logs/example.com/http/access.log |'
+ cat
+ [[ -f /home/user/example.com/awstats-dh/awstats.example.com-http.conf ]]
```

I'd expect the above to have outputted the result with HTTPS as well.

**Comment by chuckhoupt** [comment]
The processing of old logs was definitely a bug -- I've adjusted the script to only look at logs updated in the last 31 days. I also swapped the order, so HTTPS sites are processed first. Try update your install (`git pull`) and run `update-awstats` again.

The script may still fail to process all active sites. One possibility is that it is running afoul of [DH's killbot](https://help.dreamhost.com/hc/en-us/articles/216540488-Why-were-my-processes-killed-on-my-Shared-server-), especially if it takes more than a minute or so to run. This could happen if some of the sites retain more then the standard 30 days of logs, or have very large logs. By default the script directs AWStats to decompress all log files, which requires a lot of processing power.

If you think the killbot is stopping the script, then one option is to only update the last 24-48 hours of logs (this is what the suggested cron job does). Add the argument `day` to the script. I.e.:

```
$ ~/example.com/awstats-dh/update-awstats day
```

In that mode, the script only looks at the uncompressed log files (`access.log` and `access.log.0`), so it is much faster and is less likely to trigger the killbot.
